File: heap/top-k-frequent-elements.rb
require 'pry'

def top_k_frequent(nums, k)
  ans = 0
  h = Hash.new(0)

  nums.each do |num|
    h[num] += 1
  end

  heap = Heap.new
  h.each do |k,v|
    heap.insert({k => v})
  end

  k.times do
    a = heap.del_max
    ans.push(a.keys[0])
  end

  ans
end


class Heap
  attr_accessor :n, :pq

  def initialize
    @n = 0
    @pq = []
  end

  def insert(v)
    @pq[@n += 1] = v
    swim(@n)
  end

  def swim(k)
    while k > 1 && less((k / 2).floor, k)
      swap((k / 2).floor, k)
      k = k/2
    end
  end

  def swap(i, j)
    temp = @pq[i]
    @pq[i] = @pq[j]
    @pq[j] = temp
  end

  def less(i, j)
    @pq[i].values[0] < @pq[j].values[0]
  end

  def del_max
    max = @pq[1]
    swap(1, @n)
    @n -= 1
    @pq[@n + 1] = nil
    sink(1)
    max
  end

  def sink(k)
    while 2 * k <= @n
      j = 2 * k

      if !@pq[j + 1].nil?
        j += 1 if j > 1 && @pq[j].values[0] < @pq[j + 1].values[0]
      end
      break if !less(k, j)

      swap(k, j)
      k = j
    end
  end
end

top_k_frequent([5,3,1,1,1,3,73,1],2)

File: heap/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    let heap = new Heap();
    let t = {};
    let ans = []

    for(let i=0; i < nums.length; i++) {
       let num = nums[i];
       t[num] = t[num] ? t[num] + 1 : 1
    }

    for(const key in t) {
        let o = {}
        o[key] = t[key]
        heap.insert(o)
    }

    for(let i = k; k > 0; k--) {
      let el = heap.delMax();
      ans.push(Object.keys(el)[0])
    }
    return ans;
};

class Heap {
    constructor() {
        this.n = 0;
        this.pq = [];
    }

    size() {
        return this.n
    }

    isEmpty() {
        return this.n === 0;
    }

    insert(v) {
        this.pq[++this.n] = v;
        this.swim(this.n);
    }

    swim(k) {
        while(k > 1 && this.less(Math.floor(k / 2), k)) {
            this.exch(Math.floor(k / 2), k)
            k = Math.floor(k / 2)
        }
    }

    delMax() {
        let max = this.pq[1];
        this.exch(1, this.n--);
        this.pq[this.n + 1] = null;
        this.sink(1)
        return max
    }

    sink(k) {
        while(2 * k <= this.n) {
            let j = 2 * k;
            if(j + 1 <= this.n) {
                if(j > 1 && this.less(j, j + 1)) j++
            }

            if(!this.less(k,j)) break;

            this.exch(k, j);
            k = j;
        }
    }

    less(i, j) {
        return Object.values(this.pq[i])[0] < Object.values(this.pq[j])[0]
    }

    exch(i, j) {
        let temp = this.pq[i];
        this.pq[i] = this.pq[j];
        this.pq[j] = temp;
    }
}

topKFrequent([1,1,1,2,2,3], 2)
